ovn-tester/ovn_sandbox.py

```diff
@@ -1,66 +1,86 @@
 import logging
-import paramiko
+import asyncssh
 
 from ovn_exceptions import SSHError
 
 log = logging.getLogger(__name__)
 
+# asyncssh is VERY chatty when its log level is anything looser than
+# "warning".
+asyncssh.set_log_level(logging.WARNING)
+
+
+async def create_ssh(hostname, log):
+    ssh = SSH(hostname, log)
+    # XXX Need to mirror the paramiko auto add policy?
+    ssh.conn = await asyncssh.connect(hostname)
+    return ssh
+
 
 class SSH:
     def __init__(self, hostname, cmd_log):
         self.hostname = hostname
-        self.ssh = paramiko.SSHClient()
-        self.ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-        self.ssh.connect(hostname)
         self.cmd_log = cmd_log
+        self.conn = None
 
     @staticmethod
     def printable_result(out):
         if '\n' in out or '\r' in out:
             out = "---\n" + out
         return out
 
-    def run(self, cmd="", stdout=None, raise_on_error=False):
+    async def run(self, cmd="", stdout=None, raise_on_error=False):
         if self.cmd_log:
             log.info(f'Logging command: ssh {self.hostname} "{cmd}"')
 
-        ssh_stdin, ssh_stdout, ssh_stderr = self.ssh.exec_command(cmd)
-        exit_status = ssh_stdout.channel.recv_exit_status()
-
-        if exit_status != 0 and raise_on_error:
-            out = self.printable_result(ssh_stderr.read().decode().strip())
-            if len(out):
+        while True:
+            try:
+                result = await self.conn.run(cmd)
+            except asyncssh.misc.ChannelOpenError:
+                # We sometimes see these exceptions seemingly at random.
+                # Just retry if it happens.
+                continue
+            else:
+                break
+
+        if result.exit_status != 0 and raise_on_error:
+            out = self.printable_result(result.stderr)
+            if len(out) > 0:
                 log.warning(out)
             raise SSHError(
-                f'Command "{cmd}" failed with exit_status {exit_status}.'
+                f'Command {cmd} failed with exit_status {result.exit_status}.'
             )
 
-        if not ssh_stdout.channel.recv_ready():
-            return
-
         if stdout:
-            stdout.write(ssh_stdout.read().decode('ascii'))
+            stdout.write(result.stdout.strip())
         else:
-            out = self.printable_result(ssh_stdout.read().decode().strip())
-            if len(out):
+            out = self.printable_result(result.stdout.strip())
+            if len(out) > 0:
                 log.info(out)
 
 
+async def create_physical_node(hostname, log_cmds):
```

Comment:
Hi @putnopvut,
i think here we only handled one ssh session per node and that could be a bottleneck for provisioning I'm not sure if that is something that could be changed but I tried it with multisession and seems that improve the provisioning please take a look on the change below :

ovn-tester/ovn_sandbox.py
```
@@ -1,4 +1,5 @@
 import logging
+import asyncio
 import asyncssh
 
 from ovn_exceptions import SSHError
@@ -12,6 +13,7 @@ asyncssh.set_log_level(logging.WARNING)
 
 async def create_ssh(hostname, log):
     ssh = SSH(hostname, log)
     # XXX Need to mirror the paramiko auto add policy?
     ssh.conn = await asyncssh.connect(hostname)
     return ssh
@@ -65,22 +67,30 @@ class SSH:
         return await self.conn.create_process(cmd)
 
 
-async def create_physical_node(hostname, log_cmds):
-    node = PhysicalNode()
-    node.ssh = await create_ssh(hostname, log_cmds)
+async def create_physical_node(hostname, log_cmds, num_conn=1):
+    node = PhysicalNode(num_conn)
+    for _ in range(node.num_conn):
+        node.ssh.append(await create_ssh(hostname, log_cmds))
     return node
 
 
 class PhysicalNode(object):
-    def __init__(self):
-        self.ssh = None
+    def __init__(self, num_conn=1):
+        self.ssh = []
+        self.num_conn = num_conn
+        self.ssh_sem = asyncio.Semaphore(self.num_conn)
+        self.free_sessions = list(range(self.num_conn))
 
     async def run(self, cmd="", stdout=None, raise_on_error=False):
-        await self.ssh.run(cmd=cmd, stdout=stdout,
+        async with self.ssh_sem:
+            # pop is atomic ops
+            session = self.free_sessions.pop()
+            await self.ssh[session].run(cmd=cmd, stdout=stdout,
                            raise_on_error=raise_on_error)
+            self.free_sessions.append(session)
 
     async def run_deferred(self, cmd=""):
-        return await self.ssh.run_deferred(cmd=cmd)
+        return await self.ssh[0].run_deferred(cmd=cmd)
 
 
```




ovn-tester/ovn_tester.py
```
@@ -91,7 +91,7 @@ async def read_physical_deployment(deployment, global_cfg):
 
         central_dep = dep['central-node']
         central_node = await create_physical_node(
-            central_dep.get('name', 'localhost'), global_cfg.log_cmds)
+            central_dep.get('name', 'localhost'), global_cfg.log_cmds, num_conn=25)
         worker_nodes = [
             await create_physical_node(worker, global_cfg.log_cmds)
             for worker in dep['worker-nodes']
@@ -239,7 +239,6 @@ async def main(global_cfg, cluster_cfg, brex_cfg, bringup_cfg):
     central, workers = await read_physical_deployment(sys.argv[1], global_cfg)
     central_node, worker_nodes = create_nodes(cluster_cfg, central, workers)
     tests = configure_tests(config, central_node, worker_nodes)
-
     ovn = await prepare_test(central_node, worker_nodes, cluster_cfg, brex_cfg)
     await run_base_cluster_bringup(ovn, bringup_cfg)
     for test in tests:
```

Reply:
Hi Mohammad, thanks for the suggestion. My main concern with this approach is the semaphore. If there are more than `num_conn` SSH commands occurring simultaneously, then tasks will have to wait for the semaphore to become available. As the scale increases, the waits for the semaphore could result in slower test runs. Worse, if we are timing the operation, then the time spent waiting for the semaphore would factor into the reported time.

On the other hand, my current code can result in a huge number of channels being created in a single SSH connection. That also has potential not to scale well.

What if we went with something like this instead (untested. There may be syntax errors):
```
-async def create_physical_node(hostname, log_cmds):
+async def create_physical_node(hostname, log_cmds, num_conn=1):
     node = PhysicalNode()
-    node.ssh = await create_ssh(hostname, log_cmds)
+    for _ in range(num_conn):
+        node.ssh.append(await create_ssh(hostname, log_cmds))
     return node
 
 
 class PhysicalNode(object):
     def __init__(self):
-        self.ssh = None
+        self.ssh = []
+        self.cur_conn = 0
 
     async def run(self, cmd="", stdout=None, raise_on_error=False):
-        await self.ssh.run(cmd=cmd, stdout=stdout,
-                           raise_on_error=raise_on_error)
+        self.cur_conn += 1
+        if self.cur_conn >= len(self.ssh):
+            self.cur_conn = 0
+
+        await self.ssh[self.cur_conn].run(cmd=cmd, stdout=stdout,
+                                          raise_on_error=raise_on_error)
 
     async def run_deferred(self, cmd=""):
-        return await self.ssh.run_deferred(cmd=cmd)
+        return await self.ssh[0].run_deferred(cmd=cmd)
```
This uses your suggestion of having a pool of SSH connections. However, instead of performing at most one command per connection, we cycle through the connections round-robin. This makes it more likely that each connection is only using a small to medium number of channels. This also ensures that we will never wait before attempting to send an SSH command. What do you think?

Reply:
@putnopvut Hi Mark,
yes your suggestions looks very good to me :+1: